cdap-ui/graphql/helpers/BatchEndpoints/programRuns.js

```diff
@@ -18,34 +18,28 @@ const urlHelper = require('../../../server/url-helper'),
   cdapConfigurator = require('../../../server/cdap-config.js'),
   resolversCommon = require('../../resolvers-common.js');
 
+const find = require('lodash/find');
+
 let cdapConfig;
-cdapConfigurator.getCDAPConfig().then(function(value) {
+cdapConfigurator.getCDAPConfig().then(function (value) {
   cdapConfig = value;
 });
 
-async function programRecordTypeRunsResolver(programType, parent, args, context) {
-  const namespace = context.namespace;
-  const name = parent.app;
-  const programId = parent.name;
+async function batchProgramRuns(req, auth) {
+  const namespace = req[0].namespace;
   const options = resolversCommon.getPOSTRequestOptions();
   options.url = urlHelper.constructUrl(cdapConfig, `/v3/namespaces/${namespace}/runs`);
-  options.body = [
-    {
-      appId: name,
-      programType: programType,
-      programId: programId,
-    },
-  ];
-
-  const runInfo = await resolversCommon.requestPromiseWrapper(options, context.auth);
+  options.body = req.map((reqObj) => reqObj.program);
 
-  if (runInfo.length === 0) {
-    return [];
-  }
+  const runInfo = await resolversCommon.requestPromiseWrapper(options, auth);
 
-  return runInfo[0].runs;
+  return options.body.map((program) => {
+    return find(runInfo, (run) => {
```

Comment:
Won't this be efficient if we only request runs for workflow and spark programs? We know the workflow and spark programs are constant(ish). This is kind of N^2 operation here right?

Reply:
We are only requesting the runs for workflow and spark depending on the pipeline type. So essentially what's happening with this, the DataLoader requires the request array and the response array to be in the same order. However, the backend do not actually respond with the same order.

This was a quick way to solve it, but yes it's not efficient. We can generate a map of pipeline name to the info, then have this return map to just do a lookup from that map, then it's O(n)